[JENKINS-42971] GitSCMFileSystem consumes environment and expands branch name

MartinKosicky opened the pull request:
## [JENKINS-42971](https://issues.jenkins.io/browse/JENKINS-42971) Expand environment for branch names

GitSCMFileSystem now implements new SCM Builder APi that receives the Run<?,?> that can is used to expand branch name with environment variables.

MartinKosicky commented:
Hmm i wonder what the "/* Unreliable on Windows and not a platform specific test */" means in the failing test...

MarkEWaite commented:
> Hmm i wonder what the "/* Unreliable on Windows and not a platform specific test */" means in the failing test...

:) It means that the test has been unreliable in the past on Windows and the maintainers did not have time to understand why it was unreliable on Windows.  History of that is captured in the commit messages in case you want to see the sordid details.

MartinKosicky commented:
> :) It means that the test has been unreliable in the past on Windows and the maintainers did not have time to understand why it was unreliable on Windows. History of that is captured in the commit messages in case you want to see the sordid details.

Hmm so I understand that I should just ammend commit and run it again as there might be a chance I didnt break it. I was not able to debug it in my Idea for some strange reason, so I cannot tell if this test was affected by my changes

jglick commented on pom.xml:
```diff
@@ -255,12 +257,20 @@
       <id>repo.jenkins-ci.org</id>
       <url>https://repo.jenkins-ci.org/public/</url>
     </repository>
+    <repository> <!-- TODO FOR DRAFT -->
+      <id>incrementals.jenkins-ci.org</id>
+      <url>https://repo.jenkins-ci.org/incrementals/</url>
+    </repository>
```
```suggestion
```

jglick commented on pom.xml:
```diff
@@ -255,12 +257,20 @@
       <id>repo.jenkins-ci.org</id>
       <url>https://repo.jenkins-ci.org/public/</url>
     </pluginRepository>
+    <pluginRepository><!-- TODO FOR DRAFT -->
+      <id>incrementals.jenkins-ci.org</id>
+      <url>https://repo.jenkins-ci.org/incrementals/</url>
+    </pluginRepository>
```
```suggestion
```

MartinKosicky replied:
I must say that I wasnt able to compile it locally without these temporary .pom changes, but I assume they will be removed before it will be merged

MartinKosicky commented:
I did a small refactor of the calculation of the headName. I dont know what is the best practice in this project, I made a dedicated public static method, so that it could be tested standalone. Do you prefer a utility class or can we leave it as is ?

MartinKosicky commented:
I need it in incrementals, so I can make an integration test in workflow-cps

jglick commented:
> I need it in incrementals, so I can make an integration test in workflow-cps, so i converted it to to PR (removed draft)

Incrementals are published from draft PRs.

jglick commented on pom.xml:
```diff
@@ -61,7 +61,7 @@
     <changelist>-SNAPSHOT</changelist>
     <gitHubRepo>jenkinsci/${project.artifactId}-plugin</gitHubRepo>
-    <jenkins.version>2.303.3</jenkins.version>
+    <jenkins.version>2.319.3</jenkins.version>
```
Update https://github.com/jenkinsci/git-plugin/blob/d7159dc8566217e0e161fb67c8771585ade67f95/pom.xml#L78 accordingly.

MartinKosicky replied:
Ok i updated it, I hope correctly

jglick commented on src/main/java/jenkins/plugins/git/GitSCMFileSystem.java:
```diff
@@ -283,15 +284,66 @@ public boolean supportsDescriptor(SCMSourceDescriptor descriptor) {
+            private final String headName;
+            private final String prefix;
+
+            public HeadNameResult(String headName, String prefix) {
```
Stylistically I would recommend
```suggestion
            private HeadNameResult(String headName, String prefix) {
```
and then moving `calculateHeadName` inside this nested class, and renaming it accordingly (to simply `calculate`, or `create`, or `make`, or `of`, etc.).

jglick commented on src/main/java/jenkins/plugins/git/GitSCMFileSystem.java:
```diff
@@ -283,15 +284,66 @@ public boolean supportsDescriptor(SCMSourceDescriptor descriptor) {
             return AbstractGitSCMSource.class.isAssignableFrom(descriptor.clazz);
         }
 
+        public static class HeadNameResult
```
```suggestion
        static class HeadNameResult
```
Not used outside this class and the test, right?

(In which case you could also simplify a bit by dropping `private` from the fields and removing the getters—this is just an internal struct, no need for accessors.)

jglick commented on src/main/java/jenkins/plugins/git/GitSCMFileSystem.java:
```diff
@@ -283,15 +284,66 @@ public boolean supportsDescriptor(SCMSourceDescriptor descriptor) {
+            if (rev != null) {
+                headName = env.expand(rev.getHead().getName());
+            } else {
+                if (branchSpecExpandedName.startsWith(prefix)){
```
```suggestion
                if (branchSpecExpandedName.startsWith(prefix)) {
```

jglick commented on src/main/java/jenkins/plugins/git/GitSCMFileSystem.java:
```diff
@@ -283,15 +284,66 @@ public boolean supportsDescriptor(SCMSourceDescriptor descriptor) {
+            }
+
+            String prefix = Constants.R_HEADS;
+            if(branchSpecExpandedName.startsWith(Constants.R_TAGS)){
```
```suggestion
            if (branchSpecExpandedName.startsWith(Constants.R_TAGS)){
```

jglick commented on src/main/java/jenkins/plugins/git/GitSCMFileSystem.java:
```diff
@@ -283,15 +284,66 @@ public boolean supportsDescriptor(SCMSourceDescriptor descriptor) {
             if (!(scm instanceof GitSCM)) {
                 return null; // Spotbugs warns about unchecked cast without this check
             }
+
```
```suggestion
```

(gratuitous diff hunk)

jglick commented on src/main/java/jenkins/plugins/git/GitSCMFileSystem.java:
```diff
@@ -301,6 +353,13 @@ public SCMFileSystem build(@NonNull Item owner, @NonNull SCM scm, @CheckForNull
             }
+
+            EnvVars env = null;
+            if (_build != null)
+            {
```
```suggestion
            if (_build != null) {
```

jglick commented on src/test/java/jenkins/plugins/git/GitSCMFileSystemTest.java:
```diff
@@ -427,6 +429,43 @@ public void filesystem_supports_descriptor() throws Exception {
+                new EnvVars("BRANCH","dummy"));
+        assertEquals("master", result6.getHeadName());
+        assertEquals(Constants.R_HEADS, result6.getPrefix());
+
+
```
```suggestion
```

jglick commented on src/test/java/jenkins/plugins/git/GitSCMFileSystemTest.java:
```diff
@@ -427,6 +429,43 @@ public void filesystem_supports_descriptor() throws Exception {
 
+    @Issue("JENKINS-42971")
+    @Test
+    public void calculate_head_name_with_env() throws Exception
+    {
```
```suggestion
    public void calculate_head_name_with_env() throws Exception {
```

jglick commented on src/test/java/jenkins/plugins/git/GitSCMFileSystemTest.java:
```diff
@@ -427,6 +429,43 @@ public void filesystem_supports_descriptor() throws Exception {
+    {
+        GitSCMFileSystem.BuilderImpl.HeadNameResult result1 = GitSCMFileSystem.BuilderImpl.calculateHeadName(new BranchSpec("${BRANCH}"), null,
+                new EnvVars("BRANCH","master"));
+        assertEquals(result1.getHeadName(),"master");
+        assertEquals(result1.getPrefix(), Constants.R_HEADS);
```
```suggestion
        assertEquals("master", result1.getHeadName());
        assertEquals(Constants.R_HEADS, result1.getPrefix());
```

jglick commented on src/test/java/jenkins/plugins/git/GitSCMFileSystemTest.java:
```diff
@@ -427,6 +429,43 @@ public void filesystem_supports_descriptor() throws Exception {
+
+        GitSCMFileSystem.BuilderImpl.HeadNameResult result4 = GitSCMFileSystem.BuilderImpl.calculateHeadName(new BranchSpec("${BRANCH}"), null,
+                null);
+        assertEquals(result4.getHeadName(),"${BRANCH}");
+        assertEquals(result4.getPrefix(), Constants.R_HEADS);
```
```suggestion
        assertEquals("${BRANCH}", result4.getHeadName());
        assertEquals(Constants.R_HEADS, result4.getPrefix());
```

jglick commented on src/test/java/jenkins/plugins/git/GitSCMFileSystemTest.java:
```diff
@@ -427,6 +429,43 @@ public void filesystem_supports_descriptor() throws Exception {
     }
 
+    @Issue("JENKINS-42971")
+    @Test
```
May also want to assert behavior with a null env, undefined variable, etc.

MartinKosicky replied:
there is one for null env, but this is internal method, and there are many checks already in the calling method

jglick commented:
https://github.com/jenkinsci/git-plugin/pull/898#discussion_r448561267 should be investigated. Also that PR touches `GitTelescope` (whatever that is).

MartinKosicky commented:
I apolagize, I had my linter reconfigured for wrong project, now it should look better

jglick commented:
@MartinKosicky by the way when you force-push inside a PR, it breaks the ability of GitHub to display incremental diffs to reviewers. Generally I advise simply pushing commits to the branch, and leaving it to up the maintainer whether to squash-merge or true-merge at the end. (Some projects rebase-merge and so the list of commits inside the PR branch is actually critical, but this is rare in Jenkins components.)

MartinKosicky commented:
> ply pushing commits to the branch, and leaving it to up the maintainer whether to squash-merge or true-merge at the end. (Some projects rebase-merge and so the list of commits inside the PR branch is actually critical, but this is rare in Jenkins components.)

I'm sorry, I thought this was the prefered way (I read it somewhere on the internet that it's prefered in github's PRs, as this is my first contribution to a github project). I will do it only incrementally from now. Also I did some ammends because the builder was broken, in that case should I make a dummy commit to retrigger build?

MarkEWaite commented:
> I advise simply pushing commits to the branch, and leaving it to up the maintainer whether to squash-merge or true-merge at the end.

Considering the number of times that I've made mistakes visible in the history of the master branch of the git plugin, I don't think a history of mistakes and missteps by others will feel at all out of place in the git plugin.  If the submitter asks for a squash merge or if the merge history is especially "long and dirty", then I'll squash merge.

jglick commented:
> I read it somewhere on the internet that it's prefered in github's PRs

I am afraid every project has its own conventions, and preferences are certainly not consistent inside @jenkinsci which is quite decentralized.

> I did some ammends because the builder was broken, in that case should I make a dummy commit to retrigger build?

That is one way to trigger a fresh build of your own PR, yes. (People with write access to the repository can **Re-run** from the **Checks** tab. Fewer people have permission to directly **Rebuild** or **Replay** on ci.jenkins.io.)

MartinKosicky commented:
@jglick I would like to ask what is the flow here, for example when the https://github.com/jenkinsci/scm-api-plugin/pull/160 will be ok and merged, should we wait until the scm-api-plugin new version will be released (i dont know what is the time interval here, or release schedule) before marking this PR as ready for review?

jglick commented:
`scm-api` is set up for CD so release occurs automatically shortly after merging PRs with applicable labels. Ditto `workflow-cps`. I cannot speak for `git`.

MarkEWaite commented:
> should we wait until the scm-api-plugin new version will be released (i don't know what is the time interval here, or release schedule) before marking this PR as ready for review?

You can mark the PR as ready for review as soon as you feel it is ready.  No need to wait for releases of the dependent plugins so long as your pull request refers to an incremental build of the dependent plugins.  If your pull request refers to incremental builds of the dependent plugins, then I can perform testing with an install of those dependent plugins along with a build of your pull request.

MartinKosicky commented:
> [#898 (comment)](https://github.com/jenkinsci/git-plugin/pull/898#discussion_r448561267) should be investigated. Also that PR touches `GitTelescope` (whatever that is). Need to examine prior art.

Yes i saw that PR, and that is why I moved expansion at start, and also added tests to make sure that if BRANCH is */master it doesnt get converted to */*/master

MartinKosicky commented:
fixed some problem with merge conflict

jglick commented on pom.xml:
```diff
@@ -102,6 +102,7 @@
     </dependency>
     <dependency>
       <groupId>org.jenkins-ci.plugins</groupId>
+      <version>625.va_c59c48ce36f</version>  <!-- TODO: https://github.com/jenkinsci/scm-api-plugin/pull/160 -->
```
```suggestion
      <version>621.vda_a_b_055e58f7</version>  <!-- TODO until in BOM -->
```
https://github.com/jenkinsci/scm-api-plugin/releases/tag/621.vda_a_b_055e58f7

jglick commented on pom.xml:
```diff
@@ -191,6 +192,7 @@
     <dependency>
       <groupId>org.jenkins-ci.plugins</groupId>
       <artifactId>scm-api</artifactId>
+      <version>625.va_c59c48ce36f</version>  <!-- TODO: https://github.com/jenkinsci/scm-api-plugin/pull/160 -->
```
```suggestion
      <version>621.vda_a_b_055e58f7</version>  <!-- TODO until in BOM -->
```

MarkEWaite commented:
> Ok so If I understand correctly, since we change the version from incremental to that which was produced by CI/CD is that correct and can this be merged or are we going to wait for that BOM (which I know about only theoretically) ?

Since the scm-api that is now referenced in the pom is a released version of scm-api, this change could be merged.  No need to wait for scm api to appear in the bom.

The `TODO` is to remind us that once the version is included in the bom, then the `version` entry for scm-api can be removed.

MartinKosicky commented:
> > Ok so If I understand correctly, since we change the version from incremental to that which was produced by CI/CD is that correct and can this be merged or are we going to wait for that BOM (which I know about only theoretically) ?
> 
> Since the scm-api that is now referenced in the pom is a released version of scm-api, this change could be merged. No need to wait for scm api to appear in the bom.
> 
> The `TODO` is to remind us that once the version is included in the bom, then the `version` entry for scm-api can be removed.

Hmm oki so i guess ill wait for some additional feedback , because i am not able to merge this PR

MarkEWaite commented:
> Hmm oki so i guess ill wait for some additional feedback , because i am not able to merge this PR

I still need to review the PR.  I have others in my review queue ahead of this PR.  I will be unavailable next week while I take vacation.  It will be at least two weeks before I can review this PR.  If other maintainers would like to review it, they are welcome to do so.

The next release of the master branch of the git plugin will not happen for at least two weeks.  I don't want to risk releasing a new version of the plugin shortly before I disappear for a week of vacation.  It is better for the project if I am available for the week after a git plugin release in case there are surprises or other problems.

MartinKosicky commented:
> > Hmm oki so i guess ill wait for some additional feedback , because i am not able to merge this PR
> 
> I still need to review the PR. I have others in my review queue ahead of this PR. I will be unavailable next week while I take vacation. It will be at least two weeks before I can review this PR. If other maintainers would like to review it, they are welcome to do so.
> 
> The next release of the master branch of the git plugin will not happen for at least two weeks. I don't want to risk releasing a new version of the plugin shortly before I disappear for a week of vacation. It is better for the project if I am available for the week after a git plugin release in case there are surprises or other problems.

Ah oki then, i dont mean to rush anyone. Oki so I will check in 2 weeks or so, and enjoy vacation:)

MartinKosicky commented:
Just wanted to let you know that we took the incremental plugin builds for now, and put in our production jenkins, and configured all jobs to use lw checkout now,  it works perfectly

MartinKosicky commented:
Hello, this is just a friendly reminder as I am not sure if it wasn't forgotten :)

MarkEWaite commented:
> Hello, this is just a friendly reminder as I am not sure if it wasn't forgotten :)

Thanks for the reminder.  I've returned from a week of vacation, will be in the office for three days, then will be out of the office for four days.  I still have other reviews that precede this review in my queue, so it will be at least 2 weeks before I can review it.

Other maintainers are welcome to review it if they wish.

MartinKosicky commented:
Hello pinging again as this PR is opened too long, can we either finish it up somehow or probabbly add some improvements?   @MarkEWaite or @dwnusbaum or @car-roll  .

MartinKosicky commented:
Hmm

MartinKosicky commented:
Ping again (i apolagize in advance for pinging)

MarkEWaite commented:
Thanks for the ping.  I'll be busy all this weekend preparing for DevOps World and then busy all week at DevOps World.  I may be able to start the review a week or two after DevOps World

MartinKosicky commented:
> Thanks for the ping. I'll be busy all this weekend preparing for DevOps World and then busy all week at DevOps World. I may be able to start the review a week or two after DevOps World

Hello is it time now? :)

MarkEWaite commented:
> > Thanks for the ping. I'll be busy all this weekend preparing for DevOps World and then busy all week at DevOps World. I may be able to start the review a week or two after DevOps World
> 
> Hello is it time now? :)

Not yet.  I'm on vacation with family and friends and won't return until the end of this week.  I hope to start the review early next week, after some restart time as I return.

link89 commented:
Looking forward for this improvement, it suffers when working with a monorepo and not being able to turn on lightweight checkout.

MartinKosicky commented:
@MarkEWaite Hello can we like do this quick, there are only 2 files in this, one is test and one is the file. I am afraid that the build system will change soon and i will have to start over on this. Can you find some hour at least or 15 minutes to quick check it? some guys already approved it. There are also tests added. Please :)

MarkEWaite reviewed (approved):
Thanks for your patience @MartinKosicky .  Approved..

I need to perform some interactive testing of an incremental build before it is released.  I expect that release will happen later this week.

MartinKosicky commented:
Just to clarify, who will merge the branch, you after the manual test?

MarkEWaite commented:
> Just to clarify, who will merge the branch, you after the manual test?

Yes, I'll merge the branch and I'll perform the release.

MarkEWaite commented:
After building and installing the workflow-cps plugin with the change, I was able to confirm that this resolves [JENKINS-42971](https://issues.jenkins.io/browse/JENKINS-42971).  Thanks!

MartinKosicky commented:
Thanks :)
